app/ui/client/lib/UserAction.js

```diff
@@ -0,0 +1,177 @@
+
+const recordingTimeouts = {};
+const uploadingTimeouts = {};
+const typingTimeouts = {};
+
+const recordingRenews = {};
+const uploadingRenews = {};
+const typingRenews = {};
+
+const recordingUsers = new ReactiveDict();
+const uploadingUsers = new ReactiveDict();
+const typingUsers = new ReactiveDict();
+
+const selfActivity = { USER_RECORDING: false, USER_UPLOADING: false, USER_TYPING: false };
+
+const shownName = function(user) {
+	if (!user) {
+		return;
+	}
+	if (settings.get('UI_Use_Real_Name')) {
+		return user.name;
+	}
+	return user.username;
+};
+
+const stopActivity = (rid, activityType, extras) => Notifications.notifyRoom(rid, USER_ACTIVITY, shownName(Meteor.user()), false, activityType, extras);
+const startActivity = (rid, activityType, extras) => Notifications.notifyRoom(rid, USER_ACTIVITY, shownName(Meteor.user()), true, activityType, extras);
+
+function handleStreamAction(activity, performingUsers, rid, username, extras) {
+	const id = extras?.tmid ? extras.tmid : rid;
+	const users = performingUsers.get(id) || {};
+
+	if (activity === true) {
+		clearTimeout(users[username]);
+		users[username] = setTimeout(function() {
+			const u = performingUsers.get(id);
+			delete u[username];
+			performingUsers.set(id, u);
+		}, timeout);
+	} else {
+		clearTimeout(users[username]);
+		delete users[username];
+	}
+
+	performingUsers.set(id, users);
+}
+
+export const UserAction = new class {
+	constructor() {
+		Tracker.autorun(() => Session.get('openedRoom') && this.addStream(Session.get('openedRoom')));
+	}
+
+	get selfActivity() { return selfActivity; }
+
+	addStream(rid) {
+		if (rooms[rid]) {
+			return;
+		}
+		rooms[rid] = function(username, activity, activityType, extras) {
+			const user = Meteor.users.findOne(Meteor.userId(), { fields: { name: 1, username: 1 } });
+			if (username === shownName(user)) {
+				return;
+			}
+			if (activityType === USER_RECORDING) {
+				handleStreamAction(activity, recordingUsers, rid, username, extras);
+			} else if (activityType === USER_UPLOADING) {
+				handleStreamAction(activity, uploadingUsers, rid, username, extras);
+			} else if (activityType === USER_TYPING) {
+				handleStreamAction(activity, typingUsers, rid, username, extras);
+			}
+		};
+		return Notifications.onRoom(rid, USER_ACTIVITY, rooms[rid]);
+	}
+
+	startPerformingAction(rid, activityType, timeouts, renews, extras) {
+		if (renews[rid]) {
+			return;
+		}
+		selfActivity[activityType] = true;
+
+		renews[rid] = setTimeout(() => {
+			clearTimeout(renews[rid]);
+			delete renews[rid];
+		}, renew);
+
+		startActivity(rid, activityType, extras);
+		if (timeouts[rid]) {
+			clearTimeout(timeouts[rid]);
+			delete timeouts[rid];
+		}
+
+		timeouts[rid] = setTimeout(() => this.stop(rid, activityType), timeout);
+		return timeouts[rid];
+	}
+
+	start(rid, activityType, extras = {}) {
+		if (activityType === USER_RECORDING) {
+			this.startPerformingAction(rid, USER_RECORDING, recordingTimeouts, recordingRenews, extras);
```

Comment:
If you use a sinlge renew and timeout object you can eliminate these ifs and pass the activity type directly